tests/blast/test_blast_output.h

```diff
@@ -0,0 +1,814 @@
+
+        if (q == 1) // we want the second record to have no matches
+            continue;
+
+        for (int s = 0; s < 2; ++s)
+        {
+            records[q].matches.emplace_back(qIds[q], sIds[s]);
+            TBlastMatch & m = records[q].matches.back();
+            resize(rows(m.align), 2);
+            assignSource(row(m.align, 0), queries[q]);
+            assignSource(row(m.align, 1), subjects[s]);
+
+            localAlignment(m.align, seqanScheme(context.scoringScheme));
+
+            m.qStart = beginPosition(row(m.align, 0));
+            m.qEnd   = endPosition(row(m.align, 0));
+            m.sStart = beginPosition(row(m.align, 1));
+            m.sEnd   = endPosition(row(m.align, 1));
+
+            m.qLength = length(queries[q]);
+            m.sLength = length(subjects[s]);
+
+            m.qFrameShift = 1;
+            m.sFrameShift = 1;
+
+            computeAlignmentStats(m, context);
+            computeBitScore(m, context);
+            computeEValue(m, context);
+        }
+    }
+
+    // sort by bit-score
+    for (auto & r : records)
+        r.matches.sort();
+
+    test_blast_write_do(file, context, records, format, customFields, TFormat());
+}
+
+template <typename TScore, typename TString, BlastProgram p, BlastTabularSpec h>
+void test_blast_write_tabular_impl(int const format, // 0 only matches; 1 matches+headers; 2 records; 3 formattedFile
+                                   int const customFields, // 0 defaults; 1 customfields; 2 lowlevel impl
+                                   BlastIOContext<TScore, TString, p, h> & context)
+{
+    const char * tempFilename = SEQAN_TEMP_FILENAME();
+    char filenameBuffer[1000];
+    strcpy(filenameBuffer, tempFilename);
+
+    std::fstream fstream(filenameBuffer, std::ios_base::in | std::ios_base::out | std::ios_base::binary | std::ios_base::trunc);
+    SEQAN_ASSERT(fstream.is_open());
+
+    test_blast_write_record_match(fstream, format, customFields, context, BlastTabular());
+
+    std::string contents;
+    resize(contents, fstream.tellg());
+    fstream.seekg(0, std::ios::beg);
+    fstream.read(&contents[0], contents.size());
+    fstream.close();
+
+    // Generate the output string for comparison (since this includes the SeqAn version we don't want it completely
+    // hardcoded (and also to save lots of redundant code for the immense combinations of options)
+
+    std::string versionLine;
+    if (!context.legacyFormat)
+        versionLine.append("# BLASTP 2.2.26+ ");
+    else
+        versionLine.append("# BLASTP 2.2.26 ");
+    versionLine.append("[I/O Module of SeqAn-");
+    versionLine.append(std::to_string(SEQAN_VERSION_MAJOR));
+    versionLine.append(".");
+    versionLine.append(std::to_string(SEQAN_VERSION_MINOR));
+    versionLine.append(".");
+    versionLine.append(std::to_string(SEQAN_VERSION_PATCH));
+    versionLine.append(", http://www.seqan.de]\n");
+
+    std::string fieldsLine = "# Fields: ";
+    if (context.legacyFormat)
```

Comment:
Well, its the construction of the expected output. It can be done with less :spaghetti: but it will mean lots of duplication.